### PR TITLE
new PR

### PR DESCRIPTION
**Pull request by layzimayner**

### CODE REVIEW

File: src/main/java/com/example/demo/config/AppConfig.java [hunk 1]
```diff
@@ -0,0 +1,9 @@
+package com.example.demo.config;
+
+import org.springframework.context.annotation.ComponentScan;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+@ComponentScan(basePackages = "com.example.demo")
+public class AppConfig {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/config/AppConfig.java]: remove

File: src/main/java/com/example/demo/model/Book.java [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import java.math.BigDecimal;
+import lombok.Data;
+
+@Entity
+@Data
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/model/Book.java]: don't use Data for entities

File: src/main/java/com/example/demo/model/Book.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+public class Book {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/model/Book.java:17]: add `@Table`

File: src/main/java/com/example/demo/model/Book.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    private String title;
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/model/Book.java:22]: add required constraints

File: src/main/java/com/example/demo/repository/BookRepositoryImpl.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+
+import com.example.demo.model.Book;
+import java.util.List;
+import org.hibernate.Session;
+import org.hibernate.SessionFactory;
+import org.hibernate.Transaction;
+import org.hibernate.query.Query;
+import org.springframework.stereotype.Repository;
+
+@Repository
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/repository/BookRepositoryImpl.java:15]:
```suggestion
@Repository
@RequiredArgsConstructor
```

File: src/main/java/com/example/demo/repository/BookRepositoryImpl.java [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+public class BookRepositoryImpl implements BookRepository {
+    private SessionFactory sessionFactory;
+
+    public BookRepositoryImpl(SessionFactory sessionFactory) {
+        this.sessionFactory = sessionFactory;
+    }
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/repository/BookRepositoryImpl.java]:
```suggestion
    private final SessionFactory sessionFactory;
```

File: src/main/java/com/example/demo/repository/BookRepositoryImpl.java [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+            session = sessionFactory.openSession();
+            transaction = session.beginTransaction();
+            session.persist(book);
+            transaction.commit();
+            return book;
+        } catch (Exception e) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new RuntimeException("Can't save data", e);
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/repository/BookRepositoryImpl.java]: message should be more informative 
```suggestion
            throw new DataProcessingException("Can't save data", e);
```

File: src/main/java/com/example/demo/repository/BookRepositoryImpl.java [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+            }
+        }
+    }
+
+    @Override
+    public List<Book> findAll() {
+        try (Session session = sessionFactory.openSession()) {
+            Query<Book> query = session.createQuery("FROM book", Book.class);
+            return query.list();
+        }
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/repository/BookRepositoryImpl.java:57]: add catch block

File: src/main/java/com/example/demo/repository/BookRepositoryImpl.java [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+                session.close();
+            }
+        }
+    }
+
+    @Override
+    public List<Book> findAll() {
+        try (Session session = sessionFactory.openSession()) {
+            Query<Book> query = session.createQuery("FROM book", Book.class);
+            return query.list();
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/repository/BookRepositoryImpl.java]:
```suggestion
            return query.getResultList();
```

File: src/main/java/com/example/demo/service/BookServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,23 @@
+package com.example.demo.service;
+
+import com.example.demo.model.Book;
+import com.example.demo.repository.BookRepository;
+import java.util.List;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+@Service
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/service/BookServiceImpl.java:13]:
```suggestion
@Service
@RequiredArgsConstructor
```

File: src/main/java/com/example/demo/service/BookServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+public class BookServiceImpl implements BookService {
+    @Autowired
+    private BookRepository bookRepository;
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/service/BookServiceImpl.java]:
```suggestion
    private final BookRepository bookRepository;
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
Let's check task description
Add class with @SpringBootApplication annotation that will have a CommandLineRunner bean where you call BookService methods

Check IMPORTANT section
provide pom file
provide application.properties files

**Comment by layzimayner** [comment]
I created pom, application.properties and test/application.properties in last PR

File: pom.xml [hunk 12]
```diff
@@ -35,6 +35,11 @@
 			<artifactId>spring-boot-starter-data-jpa</artifactId>
 		</dependency>
 
+		<dependency>
+			<groupId>jakarta.validation</groupId>
```
Comment (Elena-Bruyako) [inline pom.xml]: remove dependency

File: src/main/java/com/example/demo/model/Book.java [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+import jakarta.persistence.Entity;
+import jakarta.persistence.GeneratedValue;
+import jakarta.persistence.GenerationType;
+import jakarta.persistence.Id;
+import jakarta.persistence.Table;
+import jakarta.validation.constraints.DecimalMin;
+import jakarta.validation.constraints.NotBlank;
+import java.math.BigDecimal;
+
+@Entity
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/model/Book.java:13]:
```suggestion
@Entity
@Getter
@Setter
```

File: src/main/java/com/example/demo/model/Book.java [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+    private String author;
+    @NotBlank(message = "ISBN is required")
+    @Column(unique = true, nullable = false)
+    private String isbn;
+    @DecimalMin(value = "0.0", inclusive = false, message = "Price must be greater than 0")
+    private BigDecimal price;
+    private String description;
+    private String coverImage;
+
+    public void setId(Long id) {
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/model/Book.java]: remove all set and get methods

File: src/main/java/com/example/demo/model/Book.java [hunk 15]
```diff
@@ -0,0 +1,59 @@
...
+import jakarta.validation.constraints.NotBlank;
+import java.math.BigDecimal;
+
+@Entity
+@Table(name = "books")
+public class Book {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @NotBlank(message = "Title is required")
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/model/Book.java]: use for all required columns

```suggestion
    @Column(nullable = false)
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

File: src/main/java/com/example/demo/model/Book.java [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+    @Column(nullable = false)
+    private String title;
+    @Column(nullable = false)
+    private String author;
+    @Column(unique = true, nullable = false)
+    private String isbn;
+    @DecimalMin(value = "0.0", inclusive = false, message = "Price must be greater than 0")
```
Comment (Elena-Bruyako) [inline src/main/java/com/example/demo/model/Book.java]:
```suggestion
    @Column(nullable = false)
```

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
one comment

**Review by Elena-Bruyako — APPROVED** [review]

**Comment by Elena-Bruyako** [comment]
Changes were already approved, after approving merge PR, create branch and implement next HW
Screenshot 2024-11-15 at 11 45 54

**Pull request closed without merging**